# AOC 2024 day 6

The Historians use their fancy device again, this time to whisk you all away to the North Pole prototype suit manufacturing lab... in the year 1518! It turns out that having direct access to history is very convenient for a group of historians.

You still have to be careful of time paradoxes, and so it will be important to avoid anyone from 1518 while The Historians search for the Chief. Unfortunately, a single guard is patrolling this part of the lab.

Maybe you can work out where the guard will go ahead of time so that The Historians can search safely?

You start by making a map (your puzzle input) of the situation. For example:

```
....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...
```

The map shows the current position of the guard with ^ (to indicate the guard is currently facing up from the perspective of the map). Any obstructions - crates, desks, alchemical reactors, etc. - are shown as #.

Lab guards in 1518 follow a very strict patrol protocol which involves repeatedly following these steps:

If there is something directly in front of you, turn right 90 degrees.
Otherwise, take a step forward.
Following the above protocol, the guard moves up several times until she reaches an obstacle (in this case, a pile of failed suit prototypes):

```
....#.....
....^....#
..........
..#.......
.......#..
..........
.#........
........#.
#.........
......#...
```

Because there is now an obstacle in front of the guard, she turns right before continuing straight in her new facing direction:

```
....#.....
........>#
..........
..#.......
.......#..
..........
.#........
........#.
#.........
......#...
```

Reaching another obstacle (a spool of several very long polymers), she turns right again and continues downward:

```
....#.....
.........#
..........
..#.......
.......#..
..........
.#......v.
........#.
#.........
......#...
```

This process continues for a while, but the guard eventually leaves the mapped area (after walking past a tank of universal solvent):

```
....#.....
.........#
..........
..#.......
.......#..
..........
.#........
........#.
#.........
......#v..
```

By predicting the guard's route, you can determine which specific positions in the lab will be in the patrol path. Including the guard's starting position, the positions visited by the guard before leaving the area are marked with an X:

```
....#.....
....XXXXX#
....X...X.
..#.X...X.
..XXXXX#X.
..X.X.X.X.
.#XXXXXXX.
.XXXXXXX#.
#XXXXXXX..
......#X..
```

In this example, the guard will visit 41 distinct positions on your map.

Predict the path of the guard. How many distinct positions will the guard visit before leaving the mapped area?

Your puzzle answer was 4647.

map:

```
....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...
```

In [3]:
!pip install -q 'advent-of-code-data[nb]'

In [5]:
import os
from aocd import get_data
inp = get_data(day=6, year=2024)

In [6]:
sample = """....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#..."""

In [7]:
def parse_map(s:str):
    return [line for line in s.splitlines()]
sample_map = parse_map(sample)
print(sample_map)

['....#.....', '.........#', '..........', '..#.......', '.......#..', '..........', '.#..^.....', '........#.', '#.........', '......#...']


In [8]:
def print_map(m):
    for line_i,line in enumerate(m):
        for x,ch in enumerate(line):
            print(f"{ch} ({x},{line_i}) ",end="")
        print()
print_map(sample_map)

. (0,0) . (1,0) . (2,0) . (3,0) # (4,0) . (5,0) . (6,0) . (7,0) . (8,0) . (9,0) 
. (0,1) . (1,1) . (2,1) . (3,1) . (4,1) . (5,1) . (6,1) . (7,1) . (8,1) # (9,1) 
. (0,2) . (1,2) . (2,2) . (3,2) . (4,2) . (5,2) . (6,2) . (7,2) . (8,2) . (9,2) 
. (0,3) . (1,3) # (2,3) . (3,3) . (4,3) . (5,3) . (6,3) . (7,3) . (8,3) . (9,3) 
. (0,4) . (1,4) . (2,4) . (3,4) . (4,4) . (5,4) . (6,4) # (7,4) . (8,4) . (9,4) 
. (0,5) . (1,5) . (2,5) . (3,5) . (4,5) . (5,5) . (6,5) . (7,5) . (8,5) . (9,5) 
. (0,6) # (1,6) . (2,6) . (3,6) ^ (4,6) . (5,6) . (6,6) . (7,6) . (8,6) . (9,6) 
. (0,7) . (1,7) . (2,7) . (3,7) . (4,7) . (5,7) . (6,7) . (7,7) # (8,7) . (9,7) 
# (0,8) . (1,8) . (2,8) . (3,8) . (4,8) . (5,8) . (6,8) . (7,8) . (8,8) . (9,8) 
. (0,9) . (1,9) . (2,9) . (3,9) . (4,9) . (5,9) # (6,9) . (7,9) . (8,9) . (9,9) 


**AI Prompt**

Remind me briefly the terse python syntax for enums, please

**AI Response**

Here's the basic enum syntax in Python:

```python
from enum import Enum

class Direction(Enum):
    NORTH = 1
    SOUTH = 2
    EAST = 3
    WEST = 4
```

You can also use `auto()` to automatically assign values:

```python
from enum import Enum, auto

class Direction(Enum):
    NORTH = auto()
    SOUTH = auto()
```

Would you like to see any specific usage examples?

**AI Prompt**

Even terser?

**AI Response**

The tersest enum syntax using string values:

```python
Direction = Enum('Direction', 'NORTH SOUTH EAST WEST')
```

Or with custom values:

```python
Direction = Enum('Direction', {'NORTH':1, 'SOUTH':2, 'EAST':3, 'WEST':4})
```

**AI Prompt**

Really, it works with just spaces separating the tokens in the first example??

**AI Response**

Yes! When you pass a space-separated string, Python's Enum constructor automatically splits it into tokens and creates enum members with auto-incrementing values starting from 1. It's equivalent to:

```python
Direction = Enum('Direction', ['NORTH', 'SOUTH', 'EAST', 'WEST'])
```

Would you like me to show you what the values would be?

In [9]:
from enum import Enum
Direction = Enum("Direction","north south east west")
Action = Enum("Action","forward turn")
Content = Enum("Content","obstruction tile exit")

In [10]:
class State:
    pos:tuple[int,int] = (0,0)
    direction:Direction = Direction.north

> **After-analysis**: This last should have been a dataclass but somehow I often forget to use them I think because of the decorator syntax makes it feel like it's not intrinsic to the language>

In [11]:
def evolve(state:State,visited:set,mmap:list[str]) -> tuple[State,set]:
    new_state = compute_next_guard_state(state,mmap)
    if content(new_state.pos,mmap) != Content.exit:
        new_visited = visited | {new_state.pos}
    else:
        new_visited = visited
    return (new_state,new_visited)

In [12]:
def compute_next_guard_state(state:State,mmap):
    pos = state.pos
    match state.direction:
        case Direction.north: facing_pos = (pos[0],pos[1]-1)
        case Direction.south: facing_pos = (pos[0],pos[1]+1)
        case Direction.east:  facing_pos = (pos[0]+1,pos[1])
        case Direction.west:  facing_pos = (pos[0]-1,pos[1])
    match content(facing_pos,mmap):
        case Content.obstruction: next_action = Action.turn
        case Content.tile: next_action = Action.forward
        case Content.exit: next_action = Action.forward
    match next_action:
        case Action.forward:
            new_pos = facing_pos
            new_direction = state.direction
        case Action.turn:
            new_pos = state.pos
            new_direction = {
                Direction.north:Direction.east,
                Direction.east:Direction.south,
                Direction.south:Direction.west,
                Direction.west:Direction.north}[state.direction]
    new_state = State()
    new_state.pos=new_pos
    new_state.direction=new_direction
    return new_state

In [13]:
def content(pos,mmap):
    if pos[1] < 0 or not pos[1] < len(mmap): return Content.exit
    if pos[0] < 0 or not pos[0] < len(mmap[0]): return Content.exit
    ch = mmap[pos[1]][pos[0]]
    if ch == '.': return Content.tile
    elif ch == '#': return Content.obstruction
    else:
        print("Found unexpected character within map bounds")
        return Content.exit

In [14]:
initial_state = State()
initial_state.pos=(3,6)
initial_state.direction=Direction.north

In [15]:
def print_map(m,visited:set|None=None,state:State|None=None):
    for line_i,line in enumerate(m):
        for x,ch in enumerate(line):
            if visited is not None and ch in visited: ch = "X"
            if state is not None and state.pos==(x,line_i):
                ch = {Direction.north:"^",Direction.east:">",Direction.west:"<",Direction.south:"v"}[state.direction]
            print(f"{ch} ({x},{line_i}) ",end="")
        print()
print_map(sample_map)

. (0,0) . (1,0) . (2,0) . (3,0) # (4,0) . (5,0) . (6,0) . (7,0) . (8,0) . (9,0) 
. (0,1) . (1,1) . (2,1) . (3,1) . (4,1) . (5,1) . (6,1) . (7,1) . (8,1) # (9,1) 
. (0,2) . (1,2) . (2,2) . (3,2) . (4,2) . (5,2) . (6,2) . (7,2) . (8,2) . (9,2) 
. (0,3) . (1,3) # (2,3) . (3,3) . (4,3) . (5,3) . (6,3) . (7,3) . (8,3) . (9,3) 
. (0,4) . (1,4) . (2,4) . (3,4) . (4,4) . (5,4) . (6,4) # (7,4) . (8,4) . (9,4) 
. (0,5) . (1,5) . (2,5) . (3,5) . (4,5) . (5,5) . (6,5) . (7,5) . (8,5) . (9,5) 
. (0,6) # (1,6) . (2,6) . (3,6) ^ (4,6) . (5,6) . (6,6) . (7,6) . (8,6) . (9,6) 
. (0,7) . (1,7) . (2,7) . (3,7) . (4,7) . (5,7) . (6,7) . (7,7) # (8,7) . (9,7) 
# (0,8) . (1,8) . (2,8) . (3,8) . (4,8) . (5,8) . (6,8) . (7,8) . (8,8) . (9,8) 
. (0,9) . (1,9) . (2,9) . (3,9) . (4,9) . (5,9) # (6,9) . (7,9) . (8,9) . (9,9) 


> **After-analysis**: In retrospect, this wasn't very useful. A nicer pattern for printing and debugging grids would be to print the coordinates or the units digit of the coordinatea long a header row and and column

> **Meta-analysis**: Maybe if this debug helper had helped more, I would not have misstepped with my analytical attempt.

In [16]:
import copy

def starting_map_state_visited(mmap):
    direction_dict = {Direction.north:"^",Direction.east:">",Direction.west:"<",Direction.south:"v"}
    g_to_dir = {v:k for k,v in direction_dict.items()}
    def f():
        for y,line in enumerate(mmap):
            for x,ch in enumerate(line):
                if ch in g_to_dir:
                    initial_pos = (x,y)
                    initial_dir = g_to_dir[ch]
                    retval = State()
                    retval.pos = initial_pos
                    retval.direction = initial_dir
                    return retval
        print("error. map had no guard on it.")
        return None
    initial_state = f()
    (x,y) = initial_state.pos
    mmap = copy.copy(mmap)
    mmap[y] = mmap[y][:x] + '.' + mmap[y][x+1:]
    visited = set()
    visited.add((x,y))
    return (initial_state, visited, mmap)

This func required a few iterations to fix typos, but nothing conceptual. The main point was to realize that it is necessary, to read the guard's initial state off of the map, and then to leave behind a clean map I can use for future movement.

In [17]:
m = parse_map(sample)
print(m)

['....#.....', '.........#', '..........', '..#.......', '.......#..', '..........', '.#..^.....', '........#.', '#.........', '......#...']


In [18]:
(init_state, init_visited, init_map) = starting_map_state_visited(m)

In [19]:
init_state.pos, init_state.direction

((4, 6), <Direction.north: 1>)

In [20]:
init_visited

{(4, 6)}

In [25]:
(init_state,init_visited,clean_map) = starting_map_state_visited(parse_map(sample))

In [26]:
evolve(init_state,init_visited,clean_map)

(<__main__.State at 0x7f3d7f6a2ea0>, {(4, 5), (4, 6)})

In [27]:
def wander(s:str,verbose=False):
    (state,visited,init_map) = starting_map_state_visited(parse_map(s))
    if verbose: print_map(m=init_map,state=state,visited=visited)
    while content(pos=state.pos,mmap=init_map) != Content.exit:
        (state,visited) = evolve(state=state,visited=visited,mmap=init_map)
        if verbose: print_map(m=init_map,state=state,visited=visited)
    return visited

In [28]:
final_visited = wander(sample)
len(final_visited)

41

In [29]:
len( wander(inp))

4647

> **Meta-analysis**: Reflections on solution method:
> 
> - I used AI mainly for lookup on odd corners of Python: enum syntax, immutable set unioning.
> - I made iterations not visible in the notebook, but they were fairly shallow.
> - Clearly, I did not write a plan for myself, but I started immediately writing types and then top-down code.
>
> However, this explicit and high-level code was so clear to me that it served as a plan, I would say. Maybe this only reflects I was comfortable with this problem so far. I could have used AI to do more work for me, but I don't think I would have understood the result as well so I doubt it would have saved time on net.

## pt 2

Find all places where an obstruction causes a loop.

While The Historians begin working around the guard's patrol route, you borrow their fancy device and step outside the lab. From the safety of a supply closet, you time travel through the last few months and record the nightly status of the lab's guard post on the walls of the closet.

Returning after what seems like only a few seconds to The Historians, they explain that the guard's patrol area is simply too large for them to safely search the lab without getting caught.

Fortunately, they are pretty sure that adding a single new obstruction won't cause a time paradox. They'd like to place the new obstruction in such a way that the guard will get stuck in a loop, making the rest of the lab safe to search.

To have the lowest chance of creating a time paradox, The Historians would like to know all of the possible positions for such an obstruction. The new obstruction can't be placed at the guard's starting position - the guard is there right now and would notice.

In the above example, there are only 6 different positions where a new obstruction would cause the guard to get stuck in a loop. The diagrams of these six situations use O to mark the new obstruction, | to show a position where the guard moves up/down, - to show a position where the guard moves left/right, and + to show a position where the guard moves both up/down and left/right.

Option one, put a printing press next to the guard's starting position:

```
....#.....
....+---+#
....|...|.
..#.|...|.
....|..#|.
....|...|.
.#.O^---+.
........#.
#.........
......#...
```

Option two, put a stack of failed suit prototypes in the bottom right quadrant of the mapped area:


```
....#.....
....+---+#
....|...|.
..#.|...|.
..+-+-+#|.
..|.|.|.|.
.#+-^-+-+.
......O.#.
#.........
......#...
```

Option three, put a crate of chimney-squeeze prototype fabric next to the standing desk in the bottom right quadrant:

```
....#.....
....+---+#
....|...|.
..#.|...|.
..+-+-+#|.
..|.|.|.|.
.#+-^-+-+.
.+----+O#.
#+----+...
......#...
```

Option four, put an alchemical retroencabulator near the bottom left corner:

```
....#.....
....+---+#
....|...|.
..#.|...|.
..+-+-+#|.
..|.|.|.|.
.#+-^-+-+.
..|...|.#.
#O+---+...
......#...
```

Option five, put the alchemical retroencabulator a bit to the right instead:

```
....#.....
....+---+#
....|...|.
..#.|...|.
..+-+-+#|.
..|.|.|.|.
.#+-^-+-+.
....|.|.#.
#..O+-+...
......#...
```

Option six, put a tank of sovereign glue right next to the tank of universal solvent:

```
....#.....
....+---+#
....|...|.
..#.|...|.
..+-+-+#|.
..|.|.|.|.
.#+-^-+-+.
.+----++#.
#+----++..
......#O..
```

It doesn't really matter what you choose to use as an obstacle so long as you and The Historians can put it into position without the guard noticing. The important thing is having enough options that you can find one that minimizes time paradoxes, and in this example, there are 6 different positions you could choose.

You need to get the guard stuck in a loop by adding a single new obstruction. How many different positions could you choose for this obstruction?

### attempt 1

a loop-causing obstruction is one where:
- guard is on a previously visited pos
- guard has orientation which is a left turn from last visit
- obstruction forces a right turn

so:
- track visit count and direction. 
- notice visit #2. check if on 2nd visit, direction is one left turn from last visit
-

> **Meta-analysis**: In retrospect this analysis was INCORRECT in that it did not include all possible conditions for a loop-causing obustrction. 
> 
> I reached this incorrect analysis by _not looking at enough examples_. I looked at only the first 2-3 examples, and did not notice that example 4 did not meet this pattern.
> 
> This error meant I went down a blind alley below, developing a function which identifies locations on the path which are only imperfectly predictive of opportunities to place an obstruction

To track visits, I will migrate the type of `visited`:
- set[Pos]
- dict[Pos:[Direction]]

In [69]:
import copy
# redefining
def starting_map_state_visited(mmap):
    direction_dict = {Direction.north:"^",Direction.east:">",Direction.west:"<",Direction.south:"v"}
    g_to_dir = {v:k for k,v in direction_dict.items()}
    def f():
        for y,line in enumerate(mmap):
            for x,ch in enumerate(line):
                if ch in g_to_dir:
                    initial_pos = (x,y)
                    initial_dir = g_to_dir[ch]
                    retval = State()
                    retval.pos = initial_pos
                    retval.direction = initial_dir
                    return retval
        print("error. map had no guard on it.")
        return None
    initial_state = f()
    (x,y) = initial_state.pos
    mmap = copy.copy(mmap)
    mmap[y] = mmap[y][:x] + '.' + mmap[y][x+1:]
    visited = dict()
    visited[(x,y)] = [initial_state.direction]
    return (initial_state, visited, mmap)

> **Meta-analysis**: Tagging redefinitions as "redefining" seems like a good compromise. It means I don't need to update every other function that depends on the existing name, I don't obscure the original implementation if I need to go back to it, and I can still run the notebook top to bottom. 

In [31]:
# redefining
def evolve(state:State,visited:dict,mmap:list[str]) -> tuple[State,set]:
    new_state = compute_next_guard_state(state,mmap)
    if content(new_state.pos,mmap) != Content.exit:
        new_visited = copy.copy(visited)
        new_visited.setdefault(new_state.pos,[]).append(new_state.direction)
    else:
        new_visited = visited
    return (new_state,new_visited)

In [32]:
final_visited = wander(sample)
len(final_visited)

41

In [33]:
final_visited

{(4, 6): [<Direction.north: 1>, <Direction.west: 4>],
 (4, 5): [<Direction.north: 1>],
 (4, 4): [<Direction.north: 1>, <Direction.east: 3>],
 (4, 3): [<Direction.north: 1>],
 (4, 2): [<Direction.north: 1>],
 (4, 1): [<Direction.north: 1>, <Direction.east: 3>],
 (5, 1): [<Direction.east: 3>],
 (6, 1): [<Direction.east: 3>],
 (7, 1): [<Direction.east: 3>],
 (8, 1): [<Direction.east: 3>, <Direction.south: 2>],
 (8, 2): [<Direction.south: 2>],
 (8, 3): [<Direction.south: 2>],
 (8, 4): [<Direction.south: 2>],
 (8, 5): [<Direction.south: 2>],
 (8, 6): [<Direction.south: 2>, <Direction.west: 4>],
 (7, 6): [<Direction.west: 4>],
 (6, 6): [<Direction.west: 4>, <Direction.south: 2>],
 (5, 6): [<Direction.west: 4>],
 (3, 6): [<Direction.west: 4>],
 (2, 6): [<Direction.west: 4>, <Direction.north: 1>],
 (2, 5): [<Direction.north: 1>],
 (2, 4): [<Direction.north: 1>, <Direction.east: 3>],
 (3, 4): [<Direction.east: 3>],
 (5, 4): [<Direction.east: 3>],
 (6, 4): [<Direction.east: 3>, <Direction.south:

In [34]:
def find_loop_chances(visited:dict):
    def has_second_visit(pos): return len(visited[pos]) > 1
    def oriented_left_since_last(pos):
        first_orientation = visited[pos][0]
        second_orientation = visited[pos][1]
        retval = (first_orientation,second_orientation) in [
            (Direction.north,Direction.west),
            (Direction.west,Direction.south),
            (Direction.south,Direction.east),
            (Direction.east,Direction.north),
            ]
        return retval
    return [pos for pos in visited.keys() if has_second_visit(pos) and oriented_left_since_last(pos)]

In [35]:
lcs = find_loop_chances(final_visited)

In [36]:
len(lcs)

3

In [37]:
lcs

[(4, 6), (6, 6), (6, 7)]

These are too few. There should be six. Do I have the right anaysis of how to find these spots?

In [38]:
# redefining
def print_map(m,visited:dict|None=None,state:State|None=None):
    for line_i,line in enumerate(m):
        for x,ch in enumerate(line):
            if visited is not None and (x,line_i) in visited: 
                ch = str( len( visited[(x,line_i)] ) )
            if state is not None and state.pos==(x,line_i):
                ch = {Direction.north:"^",Direction.east:">",Direction.west:"<",Direction.south:"v"}[state.direction]
            print(f"{ch} ({x},{line_i}) ",end="")
        print()
print_map(m,visited=final_visited)

. (0,0) . (1,0) . (2,0) . (3,0) # (4,0) . (5,0) . (6,0) . (7,0) . (8,0) . (9,0) 
. (0,1) . (1,1) . (2,1) . (3,1) 2 (4,1) 1 (5,1) 1 (6,1) 1 (7,1) 2 (8,1) # (9,1) 
. (0,2) . (1,2) . (2,2) . (3,2) 1 (4,2) . (5,2) . (6,2) . (7,2) 1 (8,2) . (9,2) 
. (0,3) . (1,3) # (2,3) . (3,3) 1 (4,3) . (5,3) . (6,3) . (7,3) 1 (8,3) . (9,3) 
. (0,4) . (1,4) 2 (2,4) 1 (3,4) 2 (4,4) 1 (5,4) 2 (6,4) # (7,4) 1 (8,4) . (9,4) 
. (0,5) . (1,5) 1 (2,5) . (3,5) 1 (4,5) . (5,5) 1 (6,5) . (7,5) 1 (8,5) . (9,5) 
. (0,6) # (1,6) 2 (2,6) 1 (3,6) 2 (4,6) 1 (5,6) 2 (6,6) 1 (7,6) 2 (8,6) . (9,6) 
. (0,7) 2 (1,7) 1 (2,7) 1 (3,7) 1 (4,7) 1 (5,7) 2 (6,7) 2 (7,7) # (8,7) . (9,7) 
# (0,8) 2 (1,8) 1 (2,8) 1 (3,8) 1 (4,8) 1 (5,8) 2 (6,8) 1 (7,8) . (8,8) . (9,8) 
. (0,9) . (1,9) . (2,9) . (3,9) . (4,9) . (5,9) # (6,9) 1 (7,9) . (8,9) . (9,9) 


My criteria are wrong. the loop blocker opportunity when guard is at 2,8 westbound is not his second time htere.

> **Meta-analysis**: Here is where I realized that my original analysis of a rule for finding places to put blockers was incorrect, because it did not produce the result I knew to be correct for the sample.
> 
> At this point, I had a choice: (1) re-commit to the broad approach of finding an analytical way to recognize "block spots" without simulating paths, or (2) start simulating paths.
>
> I chose (2), and I chose to do it in the simplest possible way, with a brute force search.
>
> Maybe what I _should_ have done was study the example cases in more detail to fortify my intuition. Why didn't I? I think, because it required awkward decoding to map coordinates to positions on a tiny ASCII grid in another part of the notebook. So I was avoiding a tedious, hi-focus data task because I was lazy perhaps?

### attempt 2: brute force

Another way to find loops: brute force. Test every position. How? Put a blocker there, and then set the gaurd wandering. and detect loops.

In [40]:
def try_wander(mmap,plus_blocker) -> bool:
    # returns true if it find_loop_chances
    pass

In [41]:
def wander_until(mmap,verbose=False):
    (state,visited,init_map) = starting_map_state_visited(mmap)
    if verbose: print_map(m=init_map,state=state,visited=visited)
    while True:
        if content(pos=state.pos,mmap=init_map) == Content.exit:
            return visited,"exit"
        elif has_looped(state,visited):
            return visited,"looped"
        else:
            (state,visited) = evolve(state=state,visited=visited,mmap=init_map)
            if verbose: print_map(m=init_map,state=state,visited=visited)
    return visited,"error"

In [42]:
def has_looped(state,visited):
    visits = visited[state.pos]
    return visits[-1] in visits[:-1]

In [43]:
def try_wander(mmap,plus_blocker_pos) -> bool:
    x,y = plus_blocker_pos
    mmap = copy.copy(mmap)
    mmap[y] = mmap[y][:x] + '#' + mmap[y][x+1:]
    (_,end_reason) = wander_until(mmap)
    return end_reason == "looped"

In [44]:
try_wander(parse_map(sample), (0,0))

False

In [45]:
try_wander(parse_map(sample), (3,6))

True

In [46]:
def count_blockspots(s):
    mm = parse_map(s)
    candidates = [(x,y) for y,line in enumerate(mm) for x,ch in enumerate(line) if ch=="."]
    print(f"candidates to try: {len(candidates)}")
    t = 0
    for pos in candidates:
        t += 1 if try_wander(mm,pos) else 0
    return t

In [47]:
count_blockspots(sample)

candidates to try: 91
trying blockspot at pos=(0, 0)
trying blockspot at pos=(1, 0)
trying blockspot at pos=(2, 0)
trying blockspot at pos=(3, 0)
trying blockspot at pos=(5, 0)
trying blockspot at pos=(6, 0)
trying blockspot at pos=(7, 0)
trying blockspot at pos=(8, 0)
trying blockspot at pos=(9, 0)
trying blockspot at pos=(0, 1)
trying blockspot at pos=(1, 1)
trying blockspot at pos=(2, 1)
trying blockspot at pos=(3, 1)
trying blockspot at pos=(4, 1)
trying blockspot at pos=(5, 1)
trying blockspot at pos=(6, 1)
trying blockspot at pos=(7, 1)
trying blockspot at pos=(8, 1)
trying blockspot at pos=(0, 2)
trying blockspot at pos=(1, 2)
trying blockspot at pos=(2, 2)
trying blockspot at pos=(3, 2)
trying blockspot at pos=(4, 2)
trying blockspot at pos=(5, 2)
trying blockspot at pos=(6, 2)
trying blockspot at pos=(7, 2)
trying blockspot at pos=(8, 2)
trying blockspot at pos=(9, 2)
trying blockspot at pos=(0, 3)
trying blockspot at pos=(1, 3)
trying blockspot at pos=(3, 3)
trying blockspot 

6

In [ ]:
count_blockspots(inp)

trying blockspot at pos=(0, 0)
trying blockspot at pos=(1, 0)
trying blockspot at pos=(2, 0)
trying blockspot at pos=(3, 0)
trying blockspot at pos=(4, 0)
trying blockspot at pos=(5, 0)
trying blockspot at pos=(6, 0)
trying blockspot at pos=(7, 0)
trying blockspot at pos=(8, 0)
trying blockspot at pos=(9, 0)
trying blockspot at pos=(10, 0)
trying blockspot at pos=(11, 0)
trying blockspot at pos=(12, 0)
trying blockspot at pos=(14, 0)
trying blockspot at pos=(15, 0)
trying blockspot at pos=(16, 0)
trying blockspot at pos=(17, 0)
trying blockspot at pos=(18, 0)
trying blockspot at pos=(19, 0)
trying blockspot at pos=(20, 0)
trying blockspot at pos=(21, 0)
trying blockspot at pos=(22, 0)
trying blockspot at pos=(23, 0)
trying blockspot at pos=(24, 0)
trying blockspot at pos=(25, 0)
trying blockspot at pos=(26, 0)
trying blockspot at pos=(27, 0)
trying blockspot at pos=(28, 0)
trying blockspot at pos=(29, 0)
trying blockspot at pos=(30, 0)
trying blockspot at pos=(31, 0)
trying blockspot a

TimeoutError: 

> **Meta-analysis**: Here I realize that my brute force (BF) approach is not performant, and I need to consider if there are ways to make it work without going back to attempt 1, which is analysis.

In [48]:
mm = parse_map(inp)

candidates = ((x,y) for y,line in enumerate(mm) for x,ch in enumerate(line) if ch==".")

In [49]:
import itertools

In [50]:
len(list(candidates))

16092

**AI Prompt**

Consider the above. Can you see the approach I am trying to use in order to count the number of locations where I can add an obstacle that produces a loop? What is going wrong? I am wondering if there is a bug in my code which causes an infinite loop, or if my approach is so ineffcient that it is running out of time.

**AI Response**

Looking at your code, your approach seems sound in principle:
1. You're trying each empty space as a potential blocker location
2. For each location, you're creating a modified map with a blocker there
3. You're simulating the guard's movement until either they exit or loop
4. You're counting how many locations cause loops

The `try_wander` function seems correct, and it works on the sample data for individual positions.

However, with 16,092 candidate positions in the full input, and each position potentially requiring many steps to determine if it causes a loop, this could indeed be very slow. The timeout error suggests this is a performance issue rather than an infinite loop.

Would you like to explore some ways to make this more efficient? For instance, we could:
1. Reduce the candidate positions by only considering those near the guard's original path
2. Optimize the loop detection
3. Add early exit conditions

Which approach would you like to explore first?

### attempt 3: brute-force, but filtered to "neighborhood" of path

> **Meta-analysis**: In retrospect, I think I was mislead by the AI's suggestion to consider positions "enar the guard's original path" rather than on the path.

In [51]:
unobstructed_path = wander_until(parse_map(inp))

In [52]:
len(unobstructed_path[0])

4647

In [53]:
def neighborhood(pos):
    diffs = [(-1,0),(1,0),(0,-1),(0,1)]
    neighbors = ((dx+pos[0],dy+pos[1]) for (dx,dy) in diffs)
    return neighbors
list(neighborhood((2,7)))

[(1, 7), (3, 7), (2, 6), (2, 8)]

In [54]:
import itertools, functools
periphery = functools.reduce(lambda x,y: x | y, (set(neighborhood(pos)) for pos in unobstructed_path[0].keys()))
len(periphery)

10405

In [55]:
def in_map(pos,mmap):
    x,y = pos
    ylen = len(mmap)
    xlen = len(mmap[0])
    return 0 <= y and y < ylen and 0 <= x and x < xlen

In [56]:
def is_dot(pos,mmap): return mmap[pos[1]][pos[0]] == "."

In [57]:
mmap = parse_map(inp)
periphery = set(p for p in periphery if in_map(p,mmap) and is_dot(p,mmap))
len(periphery)

9935

**AI Prompt**

Does the above code to define a smaller set of candidates just off the path look correct?

> **Meta-analysis**: I think this pattern of asking the AI to have a look at things might be good to do more. If I had a compiler, I'd appreciate the type-checker's feedback. Why not get AI feedback as well?

**AI Response**

Yes, your approach to reduce candidates looks good! You're:

1. Getting all positions adjacent to the guard's unobstructed path using `neighborhood()`
2. Combining these into a set with `functools.reduce`
3. Filtering to keep only valid positions that are:
   - Within map bounds (`in_map()`)
   - Currently empty spaces (`is_dot()`)

This reduced the candidates from 16,092 to 9,935 positions. While this is still a lot, it's a ~38% reduction and these positions are much more likely to actually create loops since they're near the guard's path.

Would you like to try running the loop detection with this smaller set of candidates? Or shall we look at optimizing the loop detection first?

**AI Prompt**

Do you see any big opportunitis to optimize loop detection?

**AI Response**

Looking at your current loop detection code, there are a couple of potential optimizations:

1. Currently `has_looped()` checks if the latest direction appears in previous visits to that position. However, we know from the problem description that we specifically need a left turn after a previous visit. Your earlier attempt at this logic could be incorporated.

2. We might be able to add an early exit condition - if we've visited a position more than twice without finding a loop, we probably won't find one.

3. We're creating a new copy of the visited dictionary on every step in `evolve()`. We could potentially modify this to avoid the copying overhead.

Would you like to explore any of these optimizations? Or do you see other opportunities I might have missed?

In [58]:
def find_path_periphery(s):
    mmap = parse_map(s)
    unobstructed_path = wander_until(mmap)
    def neighborhood(pos):
        diffs = [(-1,0),(1,0),(0,-1),(0,1)]
        neighbors = ((dx+pos[0],dy+pos[1]) for (dx,dy) in diffs)
        return neighbors
    import itertools, functools
    periphery = functools.reduce(lambda x,y: x | y, (set(neighborhood(pos)) for pos in unobstructed_path[0].keys()))
    def in_map(pos,mmap):
        x,y = pos
        ylen = len(mmap)
        xlen = len(mmap[0])
        return 0 <= y and y < ylen and 0 <= x and x < xlen
    def is_dot(pos,mmap): return mmap[pos[1]][pos[0]] == "."
    periphery = set(p for p in periphery if in_map(p,mmap) and is_dot(p,mmap))
    return periphery

In [59]:
periphery = find_path_periphery(inp)
len(periphery)

9935

In [89]:
# redefining: searches for blockspots in periphery of unblocked path
def count_blockspots(s):
    mm = parse_map(s)
    candidates = find_path_periphery(s)
    print(f"candidates to try: {len(candidates)}")
    t = 0
    for i,pos in enumerate(candidates):
        if i % 1000 == 0: print(f"trying blockspot candidate {i=}")
        t += 1 if try_wander(mm,pos) else 0
    return t

In [90]:
%%time
answer = count_blockspots(inp)
print(answer)

candidates to try: 9935
trying blockspot candidate i=0
trying blockspot candidate i=1000
trying blockspot candidate i=2000
trying blockspot candidate i=3000
trying blockspot candidate i=4000
trying blockspot candidate i=5000
trying blockspot candidate i=6000
trying blockspot candidate i=7000
trying blockspot candidate i=8000
trying blockspot candidate i=9000
1723
CPU times: user 7min 16s, sys: 25.1 ms, total: 7min 16s
Wall time: 7min 16s


In [62]:
answer=1723

### attempt 4: further optimizing to try placing blocks only on the path

> **Meta-analysis**: This is the analytical point which is obvious now but I missed originally: that the only place one might place an obstruction, is on the unobstructed path!

In [96]:
# redefining: searches for blockspots only in the unblocked path
def count_blockspots(s):
    mm = parse_map(s)
    # try placing blockers only on original path
    uo_path = wander_until(mm)
    candidates = set([pos for pos in uo_path[0].keys() if mm[pos[1]][pos[0]] == '.'])
    print(f"candidates to try: {len(candidates)}")
    t = 0
    for i,pos in enumerate(candidates):
        if i % 1000 == 0: print(f"at candidate {pos=}")
        t += 1 if try_wander(mm,pos) else 0
    return t

In [97]:
%%time
count_blockspots(inp)

candidates to try: 4646
at candidate pos=(67, 59)
at candidate pos=(51, 122)
at candidate pos=(20, 54)
at candidate pos=(20, 101)
at candidate pos=(84, 25)
CPU times: user 2min 6s, sys: 0 ns, total: 2min 6s
Wall time: 2min 6s


1723

### Refinement 5: use AI to generate multiprocessing code

In [91]:
%%ai
What is a simple way to use multiprocessing to speed the calculation in the last two cells? The variable mm is not mutated between calls to try_wander. Each call of the function does a defensive copy.


##### Assistant

To use multiprocessing to speed up the calculation, you can use Python's `multiprocessing` module. Here's a simple way to modify your code:

In [92]:
import multiprocessing as mp

def try_wander_wrapper(args):
    mm, pos = args
    return 1 if try_wander(mm, pos) else 0

def count_blockspots(s):
    mm = parse_map(s)
    uo_path = wander_until(mm)
    candidates = set([pos for pos in uo_path[0].keys() if mm[pos[1]][pos[0]] == '.'])
    print(f"candidates to try: {len(candidates)}")

    # Create a pool of workers
    with mp.Pool() as pool:
        # Map the try_wander function to all candidates
        results = pool.map(try_wander_wrapper, [(mm, pos) for pos in candidates])

    return sum(results)

# Run the function
#result = count_blockspots(inp)
#print(f"Result: {result}")

candidates to try: 4646
Result: 1723


This modification:

1. Creates a wrapper function `try_wander_wrapper` that unpacks the arguments for `try_wander`.
2. Uses `multiprocessing.Pool()` to create a pool of worker processes.
3. Uses `pool.map()` to apply `try_wander_wrapper` to all candidates in parallel.
4. Sums up the results to get the final count.

This approach should significantly speed up the calculation by utilizing multiple CPU cores. The exact speedup will depend on your machine's capabilities.

Note: Make sure to run this script as the main program (not in an interactive environment) for multiprocessing to work correctly on all platforms.

##### User

In [93]:
%%ai
how would I time the above?

##### Assistant

To time the execution of the multiprocessing code, you can use Python's built-in `time` module. Here's how you can modify the code to include timing:

In [94]:
import multiprocessing as mp
import time

def try_wander_wrapper(args):
    mm, pos = args
    return 1 if try_wander(mm, pos) else 0

def count_blockspots(s):
    mm = parse_map(s)
    uo_path = wander_until(mm)
    candidates = set([pos for pos in uo_path[0].keys() if mm[pos[1]][pos[0]] == '.'])
    print(f"candidates to try: {len(candidates)}")

    # Create a pool of workers
    with mp.Pool() as pool:
        # Map the try_wander function to all candidates
        results = pool.map(try_wander_wrapper, [(mm, pos) for pos in candidates])

    return sum(results)

if __name__ == '__main__':
    start_time = time.time()
    
    result = count_blockspots(inp)
    
    end_time = time.time()
    execution_time = end_time - start_time
    
    print(f"Result: {result}")
    print(f"Execution time: {execution_time:.2f} seconds")

candidates to try: 4646
Result: 1723
Execution time: 26.15 seconds


This modification:

1. Imports the `time` module.
2. Records the start time before calling `count_blockspots`.
3. Records the end time after `count_blockspots` completes.
4. Calculates and prints the execution time.

Run this script, and it will output both the result and the time taken to execute the function.

If you want to compare this with the non-multiprocessing version, you can time that as well using the same method, and then compare the execution times.

### conclusion

Analysis: I am not very happy with my solution since I suspect there is a much more efficient one. But it works.

> **meta-analysis**: I think my initial approach was clear. I think I was fairly self-aware about trying different avenues. However, I think I should have worked through all their examples more closely, and that might have speeded my way to the insight that blockspots could only ever be on the path. Also, I think when my first attempt at an analytical solution failed, I probably should have re-commited to more careful analysis than falling all the way back to BF solution
